File: 3-HARD/Python/6-optimal-binary-search-tree/proposed-solution/main.py
# Challenge: Given search frequencies for different elements, implement a dynamic programming algorithm
# to construct an optimal binary search tree minimizing the average search cost.
# Use object-oriented programming and follow the DRY principle.

from optimal_bst import OptimalBST

def main():
    keys = [10, 12, 20]
    freq = [34, 8, 50]
    optimal_bst = OptimalBST(keys, freq)
    cost = optimal_bst.optimal_cost()
    print(f"Optimal BST cost: {cost}")

if __name__ == "__main__":
    main()

File: 3-HARD/JavaScript/6-optimal-binary-search-tree/proposed-solution/main.js
/*
Challenge:
Given search frequencies for different elements, implement a dynamic programming algorithm to construct an optimal binary search tree minimizing the average search cost.

This solution follows DRY principles and is implemented in JavaScript.
*/

import OptimalBST from "./OptimalBST.js";

function main() {
  const keys = [10, 12, 20];
  const freq = [34, 8, 50];

  const optimalBST = new OptimalBST(keys, freq);
  const minCost = optimalBST.optimalCost();

  console.log("Minimum cost of Optimal BST:", minCost);
}

main();
